File: internal/lsp/markdown_test.go
package lsp

import (
	"strings"
	"testing"

	"lore/internal/lore"
)

// findNodes returns every node in the tree whose Kind matches `kind`,
// walking children recursively. Test-only helper.
func findNodes(nodes []MarkdownNode, kind string) []MarkdownNode {
	var out []MarkdownNode
	for _, n := range nodes {
		if n.Kind == kind {
			out = append(out, n)
		}
		out = append(out, findNodes(n.Children, kind)...)
	}
	return out
}

// joinedText concatenates every text-leaf segment in the tree, in
// source order. Useful for assertions about content survival.
func joinedText(nodes []MarkdownNode) string {
	var b strings.Builder
	walkText(nodes, &b)
	return b.String()
}

func walkText(nodes []MarkdownNode, b *strings.Builder) {
	for _, n := range nodes {
		switch n.Kind {
		case "text":
			for _, s := range n.Segments {
				b.WriteString(s.Text)
			}
		case "code-block", "code-inline":
			b.WriteString(n.Text)
		}
		if len(n.Children) > 0 {
			walkText(n.Children, b)
		}
	}
}

func TestBuildDescriptionContentParagraph(t *testing.T) {
	got := buildDescriptionContent(lore.NewWorld(), "Ranger of the North.")
	if len(got) != 1 || got[0].Kind != "paragraph" {
		t.Fatalf("got %+v", got)
	}
	if joinedText(got) != "Ranger of the North." {
		t.Fatalf("text = %q", joinedText(got))
	}
}

func TestBuildDescriptionContentBlockquoteSpansParagraphs(t *testing.T) {
	src := "> Lots of prose.\n>\n> Spread over many lines."
	got := buildDescriptionContent(lore.NewWorld(), src)
	bqs := findNodes(got, "blockquote")
	if len(bqs) != 1 {
		t.Fatalf("expected one blockquote, got %d (tree=%+v)", len(bqs), got)
	}
	// The blockquote body keeps both halves of the quote separated by
	// the blank `>` line. CommonMark parses these as two paragraphs.
	paras := findNodes(bqs, "paragraph")
	if len(paras) != 2 {
		t.Fatalf("expected two paragraphs inside blockquote, got %d", len(paras))
	}
	text := joinedText(got)
	if !strings.Contains(text, "Lots of prose.") || !strings.Contains(text, "Spread over many lines.") {
		t.Fatalf("blockquote content lost: %q", text)
	}
	// The leading `>` markers must not survive into rendered text.
	if strings.Contains(text, ">") {
		t.Fatalf("blockquote `>` marker leaked: %q", text)
	}
}

func TestBuildDescriptionContentHorizontalRule(t *testing.T) {
	got := buildDescriptionContent(lore.NewWorld(), "before\n\n---\n\nafter")
	hrs := findNodes(got, "hr")
	if len(hrs) != 1 {
		t.Fatalf("expected one hr, got %d", len(hrs))
	}
}

func TestBuildDescriptionContentList(t *testing.T) {
	src := "- one\n- two\n- three"
	got := buildDescriptionContent(lore.NewWorld(), src)
	lists := findNodes(got, "list")
	if len(lists) != 1 || lists[0].Ordered {
		t.Fatalf("expected one unordered list, got %+v", lists)
	}
	items := findNodes(lists, "list-item")
	if len(items) != 3 {
		t.Fatalf("expected 3 items, got %d", len(items))
	}
}

func TestBuildDescriptionContentOrderedList(t *testing.T) {
	got := buildDescriptionContent(lore.NewWorld(), "1. first\n2. second")
	lists := findNodes(got, "list")
	if len(lists) != 1 || !lists[0].Ordered {
		t.Fatalf("expected one ordered list, got %+v", lists)
	}
}

func TestBuildDescriptionContentHeading(t *testing.T) {
	got := buildDescriptionContent(lore.NewWorld(), "## Subsection\n\nbody")
	headings := findNodes(got, "heading")
	if len(headings) != 1 || headings[0].Level != 2 {
		t.Fatalf("expected one h2, got %+v", headings)
	}
}

func TestBuildDescriptionContentFencedCode(t *testing.T) {
	src := "```go\nfunc f() {}\n```"
	got := buildDescriptionContent(lore.NewWorld(), src)
	codes := findNodes(got, "code-block")
	if len(codes) != 1 {
		t.Fatalf("expected one code block, got %d", len(codes))
	}
	if codes[0].Lang != "go" {
		t.Fatalf("lang = %q", codes[0].Lang)
	}
	if !strings.Contains(codes[0].Text, "func f() {}") {
		t.Fatalf("code text = %q", codes[0].Text)
	}
}

func TestBuildDescriptionContentEmphasisStrongCodeLink(t *testing.T) {
	src := "see *italic*, **bold**, `code`, and [text](https://example.com)"
	got := buildDescriptionContent(lore.NewWorld(), src)
	if len(findNodes(got, "emphasis")) == 0 {
		t.Errorf("missing emphasis")
	}
	if len(findNodes(got, "strong")) == 0 {
		t.Errorf("missing strong")
	}
	if len(findNodes(got, "code-inline")) == 0 {
		t.Errorf("missing code-inline")
	}
	links := findNodes(got, "link")
	if len(links) != 1 || links[0].Href != "https://example.com" {
		t.Fatalf("link = %+v", links)
	}
}

func TestBuildDescriptionContentEntityColouringInsideMarkdown(t *testing.T) {
	// An entity mention inside a blockquote body must end up in a text
	// leaf whose segment carries the entity's colourIndex, not as plain
	// text. Same expectation as the legacy flat-segments path.
	world := lore.NewWorld()
	world.Entities = []lore.Entity{{Name: "Aragorn", Type: "character"}}
	world.Match = lore.BuildMatchIndex(world)
	got := buildDescriptionContent(world, "> Met Aragorn at dusk.")
	var sawColoured bool
	var visit func(nodes []MarkdownNode)
	visit = func(nodes []MarkdownNode) {
		for _, n := range nodes {
			for _, s := range n.Segments {
				if strings.Contains(s.Text, "Aragorn") && s.ColourIndex >= 0 {
					sawColoured = true
				}
			}
			visit(n.Children)
		}
	}
	visit(got)
	if !sawColoured {
		t.Fatalf("expected coloured Aragorn segment, got tree: %+v", got)
	}
}

func TestBuildDescriptionContentJournalExample(t *testing.T) {
	// End-to-end shape check for the block-aside body the user
	// originally asked for: a journal item with a multi-paragraph
	// blockquote.
	src := "> Lots of prose.\n>\n> Spread over many lines."
	got := buildDescriptionContent(lore.NewWorld(), src)
	if len(findNodes(got, "blockquote")) != 1 {
		t.Fatalf("blockquote missing: %+v", got)
	}
	if joinedText(got) == "" {
		t.Fatalf("no text survived")
	}
}

File: vscode-lore/test/markdown.test.ts
import "./setup-dom.ts";
import { describe, it, beforeEach } from "node:test";
import assert from "node:assert/strict";
import {
  renderMarkdownTree,
  flattenSegments,
  type MarkdownNode,
} from "../media/wiki/markdown.ts";
import type { ContextSegment } from "../media/wiki/segments.ts";
import { resetDom } from "./setup-dom.ts";

const PALETTE = ["#ff0000", "#00ff00", "#0000ff"];

// txt is a tiny helper for a "text" leaf carrying a single plain
// segment, used to keep the test trees readable.
function txt(text: string): MarkdownNode {
  return { kind: "text", segments: [{ text }] };
}

describe("renderMarkdownTree", () => {
  beforeEach(() => resetDom());

  it("renders a paragraph as <p>", () => {
    const parent = document.createElement("div");
    renderMarkdownTree(
      [{ kind: "paragraph", children: [txt("Ranger of the North.")] }],
      parent,
      PALETTE,
      () => {},
      true,
    );
    const p = parent.querySelector("p");
    assert.ok(p, "expected a <p>");
    assert.equal(p!.textContent, "Ranger of the North.");
  });

  it("renders hr as <hr>", () => {
    const parent = document.createElement("div");
    renderMarkdownTree([{ kind: "hr" }], parent, PALETTE, () => {}, true);
    assert.ok(parent.querySelector("hr"));
  });

  it("renders blockquote with nested paragraph", () => {
    const parent = document.createElement("div");
    renderMarkdownTree(
      [{
        kind: "blockquote",
        children: [{ kind: "paragraph", children: [txt("Lots of prose.")] }],
      }],
      parent,
      PALETTE,
      () => {},
      true,
    );
    const bq = parent.querySelector("blockquote");
    assert.ok(bq, "expected a <blockquote>");
    const text = bq!.textContent ?? "";
    assert.equal(text.includes(">"), false, "`>` leaked into rendered text: " + text);
    assert.match(text, /Lots of prose/);
  });

  it("renders headings starting at <h2> (h1 reserved for entity name)", () => {
    const parent = document.createElement("div");
    renderMarkdownTree(
      [
        { kind: "heading", level: 1, children: [txt("Top")] },
        { kind: "heading", level: 2, children: [txt("Sub")] },
      ],
      parent,
      PALETTE,
      () => {},
      true,
    );
    assert.equal(parent.querySelector("h2")?.textContent, "Top");
    assert.equal(parent.querySelector("h3")?.textContent, "Sub");
  });

  it("renders unordered and ordered lists", () => {
    const parent = document.createElement("div");
    renderMarkdownTree(
      [
        {
          kind: "list",
          ordered: false,
          children: [
            { kind: "list-item", children: [{ kind: "paragraph", children: [txt("a")] }] },
            { kind: "list-item", children: [{ kind: "paragraph", children: [txt("b")] }] },
          ],
        },
        {
          kind: "list",
          ordered: true,
          children: [
            { kind: "list-item", children: [{ kind: "paragraph", children: [txt("one")] }] },
          ],
        },
      ],
      parent,
      PALETTE,
      () => {},
      true,
    );
    assert.equal(parent.querySelectorAll("ul li").length, 2);
    assert.equal(parent.querySelectorAll("ol li").length, 1);
  });

  it("renders fenced code blocks via <pre><code>", () => {
    const parent = document.createElement("div");
    renderMarkdownTree(
      [{ kind: "code-block", lang: "go", text: "func f() {}\n" }],
      parent,
      PALETTE,
      () => {},
      true,
    );
    const code = parent.querySelector("pre > code");
    assert.ok(code, "expected <pre><code>");
    assert.equal(code!.textContent, "func f() {}\n");
    assert.match(code!.className, /lang-go/);
  });

  it("renders inline emphasis, strong, code, link", () => {
    const parent = document.createElement("div");
    renderMarkdownTree(
      [{
        kind: "paragraph",
        children: [
          txt("see "),
          { kind: "emphasis", children: [txt("italic")] },
          txt(" and "),
          { kind: "strong", children: [txt("bold")] },
          txt(" plus "),
          { kind: "code-inline", text: "code" },
          txt(" and "),
          { kind: "link", href: "https://example.com", children: [txt("link")] },
        ],
      }],
      parent,
      PALETTE,
      () => {},
      true,
    );
    assert.equal(parent.querySelector("em")?.textContent, "italic");
    assert.equal(parent.querySelector("strong")?.textContent, "bold");
    assert.equal(parent.querySelector("code")?.textContent, "code");
    const a = parent.querySelector("a");
    assert.ok(a);
    assert.equal(a!.getAttribute("href"), "https://example.com");
    assert.equal(a!.textContent, "link");
  });

  it("preserves entity colour spans inside paragraph text leaves", () => {
    const parent = document.createElement("div");
    const segments: ContextSegment[] = [
      { text: "Met " },
      { text: "Aragorn", entity: "Aragorn", colourIndex: 0 },
      { text: " at dusk." },
    ];
    renderMarkdownTree(
      [{ kind: "paragraph", children: [{ kind: "text", segments }] }],
      parent,
      PALETTE,
      () => {},
      true,
    );
    const span = parent.querySelector("p > span");
    assert.ok(span, "expected entity-colour span inside paragraph");
    assert.equal(span!.textContent, "Aragorn");
  });

  it("preserves entity colour spans inside blockquote", () => {
    const parent = document.createElement("div");
    const segments: ContextSegment[] = [
      { text: "Met " },
      { text: "Aragorn", entity: "Aragorn", colourIndex: 0 },
      { text: " at dusk." },
    ];
    renderMarkdownTree(
      [{
        kind: "blockquote",
        children: [{ kind: "paragraph", children: [{ kind: "text", segments }] }],
      }],
      parent,
      PALETTE,
      () => {},
      true,
    );
    const span = parent.querySelector("blockquote span");
    assert.ok(span, "expected entity-colour span inside blockquote");
    assert.equal(span!.textContent, "Aragorn");
  });

  it("never executes embedded HTML — user prose is text only", () => {
    // A `<script>` smuggled in via a text leaf must not land as a real
    // <script> element. createElement / text nodes only — angle
    // brackets stay as literal characters.
    const parent = document.createElement("div");
    renderMarkdownTree(
      [{
        kind: "paragraph",
        children: [txt("<script>alert(1)</script>")],
      }],
      parent,
      PALETTE,
      () => {},
      true,
    );
    assert.equal(parent.querySelector("script"), null, "<script> leaked into DOM");
    assert.match(parent.textContent ?? "", /<script>alert\(1\)<\/script>/);
  });
});

describe("flattenSegments", () => {
  it("collects text-leaf segments in source order", () => {
    const out = flattenSegments([
      { kind: "paragraph", children: [
        { kind: "text", segments: [{ text: "alpha " }] },
        { kind: "strong", children: [{ kind: "text", segments: [{ text: "beta" }] }] },
        { kind: "text", segments: [{ text: " gamma." }] },
      ] },
    ]);
    assert.deepEqual(out.map((s) => s.text), ["alpha ", "beta", " gamma."]);
  });

  it("includes code spans and code blocks as plain text", () => {
    const out = flattenSegments([
      { kind: "paragraph", children: [
        { kind: "text", segments: [{ text: "before " }] },
        { kind: "code-inline", text: "X" },
        { kind: "text", segments: [{ text: " after" }] },
      ] },
      { kind: "code-block", text: "fn() {}\n" },
    ]);
    assert.deepEqual(out.map((s) => s.text), ["before ", "X", " after", "fn() {}\n"]);
  });
});
